Implement baseline QLit evaluation (ZeroQubit, OneQubit, QubitTilt, UniformSuperpos, QubitTensor, QubitUnit)

glee366 opened the pull request:
Implementation for QLit, please let me know if there are any issues.

glee366 committed:
Implement baseline QLit evaluation (ZeroQubit, OneQubit, Tilt, Superpos, Tensor)

ausbin commented on qwerty_ast/src/ast.rs:
Can you put this file back to how it was? Try this:

```
git checkout rust-ast -- qwerty_ast/src/ast.rs
```

I would be amazed if this code builds because a bunch of code expects the new structure (which is being undone here)

ausbin commented on qwerty_ast/src/repl.rs:
```diff
@@ -66,11 +66,60 @@ impl Expr {
+                    Some(Expr::QubitRef { index: q })
+                }
+                QLit::QubitTilt { q, angle_deg, dbg } => {
+                    let inside_expr = Expr::QLit {
```
Wow good point. Don't know why I didn't think of this.

Evaluating a QLit should never produce an expression. In fact, every QLit will evaluate to either a QubitRef (or a QubitTensor of QubitRefs). I proved that somewhere. So while I like your idea to do this, I can imagine it becoming verbose. Instead, we can have a separate pair of `eval_step()` and `eval_to_value()` functions just for QLits. Think of it like this inner `match` as its own function that takes and returns a `QLit`

glee366 replied:
I created a impl qlit

ausbin commented on qwerty_ast/src/repl.rs:
```diff
@@ -66,11 +66,60 @@ impl Expr {
+                        dbg: dbg.clone(),
+                    };
+                    if let Some(Expr::QubitRef { index }) = inside_expr.eval_step(state) {
+                        state.sim.h(index);
```
Does this behave how we expect?

If you're not confident, I would just leave the `UniformVectorSuperpos` arm as a `todo!()` for now since it is the trickiest

glee366 replied:
made it a todo

ausbin commented on qwerty_ast/src/repl.rs:
I think this is already becoming intricate enough that we need some tests for this. I think we would call `eval_to_value` on an expression and check the simulator state and returned expression for correctness. Let me know if I can help

glee366 replied:
only wrote tests for 0/1 qubit, next step is changing convert_ast i believe to be able to run tests for tilt and tensor?

ausbin reviewed (changes requested):
Some suggestions

glee366 committed:
added one qubit tests (copied from zero qubit), created new impl block for QLit, need to test Tilt / Tensor.

ausbin committed:
Get code compiling

ausbin committed:
`QLit::eval_step_qubit()` → `QLit::eval_step()`

ausbin committed:
Run rustfmt

ausbin committed:
Add some example unit tests

The last test fails, so (unless there is a typo in the test), the code
it is testing needs to be fixed.

ausbin commented:
We'll have to revisit this

cc @vivek-vishwanath
